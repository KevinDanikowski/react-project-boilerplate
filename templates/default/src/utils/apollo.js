// this file comes from https://github.com/zeit/next.js/blob/canary/examples/with-apollo/lib/apollo.js
import React from 'react'
import Head from 'next/head'
import { ApolloProvider } from '@apollo/react-hooks'
import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { HttpLink } from 'apollo-link-http'
import fetch from 'isomorphic-unfetch'
import getConfig from 'next/config'

const {publicRuntimeConfig} = getConfig()
const {urls} = publicRuntimeConfig
let apolloClient = null


/**
 * Creates and configures the ApolloClient
 * @param  {Object} [initialState={}]
 */
function createApolloClient(initialState = {}) {
  return new ApolloClient({
    ssrMode: typeof window === 'undefined', // Disables forceFetch on the server (so queries are only run once)
    link: new HttpLink({
      uri: urls.apiGatewayUrl, // Server URL (must be absolute)
      // credentials: 'same-origin', // Additional fetch() options like `credentials` or `headers`
      fetch,
    }),
    cache: new InMemoryCache().restore(initialState),
  })
}

/**
 * Always creates a new apollo client on the server
 * Creates or reuses apollo client in the browser.
 * @param  {Object} initialState
 */
function initApolloClient(initialState) {
  // Make sure to create a new client for every server-side request so that data
  // isn't shared between connections (which would be bad)
  if (typeof window === 'undefined') {
    return createApolloClient(initialState)
  }

  // Reuse client on the client-side
  if (!apolloClient) {
    apolloClient = createApolloClient(initialState)
  }

  return apolloClient
}

/**
 * Creates and provides the apolloContext
 * to a next.js PageTree. Use it by wrapping
 * your PageComponent via HOC pattern.
 * Optimize not using SSR by ssr = false 
 * https://nextjs.org/blog/next-9#automatic-static-optimization
 * @param {Function|Class} PageComponent
 * @param {Object} [config]
 * @param {Boolean} [config.ssr=true]
 */
export function withApollo(PageComponent, { ssr = true } = {}) {
  // eslint-disable-next-line
  const WithApollo = ({ apolloClient, apolloState, ...pageProps }) => {
    const client = apolloClient || initApolloClient(apolloState)
    return (
      // eslint-disable-next-line
      <ApolloProvider client={client}>
        <PageComponent {...pageProps} />
      </ApolloProvider>
    )
  }

  // Set the correct displayName in development
  if (process.env.NODE_ENV !== 'production') {
    const displayName =
      PageComponent.displayName || PageComponent.name || 'Component'

    if (displayName === 'App') {
      console.warn('This withApollo HOC only works with PageComponents.')
    }

    WithApollo.displayName = `withApollo(${displayName})`
  }

  if (ssr || PageComponent.getInitialProps) {
    WithApollo.getInitialProps = async ctx => {
      const { AppTree } = ctx

      // Initialize ApolloClient, add it to the ctx object so
      // we can use it in `PageComponent.getInitialProp`.
      // eslint-disable-next-line
      const apolloClient = (ctx.apolloClient = initApolloClient())

      // Run wrapped getInitialProps methods
      let pageProps = {}
      if (PageComponent.getInitialProps) {
        pageProps = await PageComponent.getInitialProps(ctx)
      }

      // Only on the server:
      if (typeof window === 'undefined') {
        // When redirecting, the response is finished.
        // No point in continuing to render
        if (ctx.res && ctx.res.finished) {
          return pageProps
        }

        // Only if ssr is enabled
        if (ssr) {
          try {
            // Run all GraphQL queries through tree
            const { getDataFromTree } = await import('@apollo/react-ssr')

            // Currently there are issues with getDataFromTree not running 
            // react hook queries on the server. Currently not a concern.
            // Some research found here: https://stackoverflow.com/questions/58021561/server-side-render-data-when-using-apollo-client-with-nextjs/59290188#59290188
            // another possible solution is put into getInitialProps until getDataFromTree supports hooks
            // Another method was https://github.com/trojanowski/react-apollo-hooks/issues/132 which didn't work
            await getDataFromTree(
              <AppTree
                pageProps={{
                  ...pageProps,
                  apolloClient,
                }}
              />
            )
          } catch (error) {
            // Prevent Apollo Client GraphQL errors from crashing SSR.
            // Handle them in components via the data.error prop:
            // https://www.apollographql.com/docs/react/api/react-apollo.html#graphql-query-data-error
            console.error('Error while running `getDataFromTree`', error)
          }

          // getDataFromTree does not call componentWillUnmount
          // head side effect therefore need to be cleared manually
          Head.rewind()
        }
      }

      // Extract query data from the Apollo store
      const apolloState = apolloClient.cache.extract()

      return {
        ...pageProps,
        apolloState,
      }
    }
  }

  return WithApollo
}

export default withApollo
