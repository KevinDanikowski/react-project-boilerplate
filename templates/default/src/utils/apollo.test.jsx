import React, { Component } from 'react'
import { render } from '@testing-library/react'
import { withApollo } from './apollo'

jest.mock('next/config', () => () => ({
  publicRuntimeConfig: {
    apiGatewayUrl: process.env.API_GATEWAY_URL
  }
}))

describe('apollo', () => {
  // eslint-disable-next-line
  class SampleApp extends Component {
    render() {
      return <div>OK!</div>
    }
  }

  it('should run without error', () => {
    const SampleComponentWithApollo = withApollo(SampleApp, { ssr: true })
    const { asFragment } = render(<SampleComponentWithApollo />)

    expect(asFragment()).toMatchSnapshot()
  })
})
