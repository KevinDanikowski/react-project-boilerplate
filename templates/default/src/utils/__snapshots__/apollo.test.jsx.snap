// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apollo should run without error 1`] = `
<DocumentFragment>
  <div>
    OK!
  </div>
</DocumentFragment>
`;
